### ISSUE TITLE
Compatibility issues in unzLocateFile

### ISSUE THREAD

**Issue by ljavorsk** [zlib-ng/minizip-ng#745]
Hi,

I've run into a building issue with the `libdigidocpp` package when switching from minizip (madler/zlib part) to minizip-ng (built with MZ_COMPAT flag).

I reported this to their upstream [1], but now when I think about it, it may be the issue of minizip-ng as it should be fully compatible, but it's not.

Could you please take a look at the error and help me resolve this building issue? I was trying to find some sort of PORTING document in your Github, but there is nothing that would look like it.
Thank you

[1] https://github.com/open-eid/libdigidocpp/issues/569

**Comment by ljavorsk** [comment]
I've added the `#include <zlib.h>` to the `ZipSerialize.cpp` file, as they were counting on minizip to include it.

But now I've encountered another error, which may be the incompatible part of minizip-ng function:
Minizip-ng definition: `unzLocateFile(unzFile file, const char *filename, unzFileNameComparer filename_compare_func)`
Minizip definition: `unzLocateFile OF((unzFile file, const char *szFileName, int iCaseSensitivity))`

Error:
```
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp:139:58: error: invalid conversion from ‘int’ to ‘unzFileNameComparer’ {aka ‘int (*)(void*, const char*, const char*)’} [-fpermissive]
  139 |     int unzResult = unzLocateFile(d->open, file.c_str(), 1);
      |                                                          ^
      |                                                          |
      |                                                          int
In file included from /usr/include/minizip/unzip.h:10,
                 from /builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp:27:
/usr/include/minizip/mz_compat.h:352:87: note:   initializing argument 3 of ‘int unzLocateFile(unzFile, const char*, unzFileNameComparer)’
  352 | ZEXPORT int     unzLocateFile(unzFile file, const char *filename, unzFileNameComparer filename_compare_func);
      |                                                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp: In member function ‘void digidoc::ZipSerialize::addFile(const std::string&, std::istream&, const Properties&, Flags)’:
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp:189:17: error: cannot convert ‘<brace-enclosed initializer list>’ to ‘uint32_t’ {aka ‘unsigned int’} in initialization
  189 |         0, 0, 0 };
      |                 ^
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp: In member function ‘digidoc::ZipSerialize::Properties digidoc::ZipSerialize::properties(const std::string&) const’:
/builddir/build/BUILD/libdigidocpp-3.16.0/src/util/ZipSerialize.cpp:225:58: error: invalid conversion from ‘int’ to ‘unzFileNameComparer’ {aka ‘int (*)(void*, const char*, const char*)’} [-fpermissive]
  225 |     int unzResult = unzLocateFile(d->open, file.c_str(), 1);
      |                                                          ^
      |                                                          |
      |                                                          int
/usr/include/minizip/mz_compat.h:352:87: note:   initializing argument 3 of ‘int unzLocateFile(unzFile, const char*, unzFileNameComparer)’
  352 | ZEXPORT int     unzLocateFile(unzFile file, const char *filename, unzFileNameComparer filename_compare_func);
      |                                                                   ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
gmake[2]: *** [src/CMakeFiles/digidocpp.dir/build.make:489: src/CMakeFiles/digidocpp.dir/util/ZipSerialize.cpp.o] Error 1
```

**Comment by nmoinvaz** [comment]
It appears they are using a [another version](https://github.com/zlib-ng/minizip-ng/blob/252588f6a78bd7656f4e03d1b236695b3182ff7d/unzip.h#L242) of `minizip` that was possibly made by myself. It was not API compatible hence the differences in the `unzLocateFile` callback. The `dev` branch is compatible with the version of `minizip` in zlib distribution.

**Comment by Neustradamus** [comment]
@nmoinvaz: Have you seen the @tuliom comment?

**Comment by nmoinvaz** [comment]
zlib distribution is https://github.com/madler/zlib.

Hmm, that does look like an incompatibility. Feel free to please submit a PR to fix it.

**Comment by tuliom** [comment]
FYI: I can't close this bug, but it has already been fixed by #746 .
